### DISCUSSION TITLE
Vitoconnect passthrough

### DISCUSSION THREAD

**Discussion by challo2018** [bertmelis/VitoWiFi#138] [Ideas]
Is there any possibilty or any user which used VItoWifi together with viesmann vitoconnect due to longer viessmann guarantee when using vitoconnect?

ESP32 DEV board has CP2102 Serial Chip which is similar to the one from original optolink kabel. 
So it is possible to connect ESP32 USP directl to the vitoconnect. But It's necessary to extend VItoWifi to put all trafic from IR Serial port to the USB Serial port and other way round.

I'm waiting for my ESP32 Dev boards to try such an implementation. Is there any user who allready tried?

**Comment by bertmelis** [comment]
Currently, VitoWiFi doesn't support sniffing.

VitoWiFi has long been neglected from my side since I didn't use it myself anymore. However, in a couple of months I will move back to a house where I have full control over the (Viessmann) heating and breath new life into this project.

I'm already thinking about creating a version 3 with more flexibility on how to use the library. Suggestions are welcome of course. Sniffing is on the list as extended warranty is definitely something you don't want to sacrifice on expensive equipment.

**Comment by bertmelis** [comment]
v3 is more modular and a sniffer is probably within the possibilities to create. not top priority...

**Comment by challo2018** [comment]
Thanks for your answer. Do you have time schedule for V3?

**Comment by bertmelis** [comment]
V3 will be very soon. The sniffer features a bit later.

I'm already testing (on a raspberry pi, code is compatible).

**Comment by challo2018** [comment]
Hello bert, any news here? I read that V3 is allready running stable on your system, so i will try it with my heat pump. do you recommend using esp32 or raspberry?

**Comment by bertmelis** [comment]
I'm running V3. I'm running on a raspberry pi because it's easier for me for testing. An ESP would be more stable but that has more to do with stability of the system (my RPI sometimes fails because of my power stability).

I'm working on a ESP version that Serial via USB.

Passthrough should also be possible with the separate components of the library. I will think of an example and add it to the repo.

**Comment by challo2018** [comment]
thank you, would be great to test such an version!

**Comment by bertmelis** [comment]
Just for my understanding:
- you connect the ESP to VitoConnect
- you connect a diy serial optolink via one of the other UARTs
- the firmware forwards USB to optolink and vice versa.

Could we start with just the forwarder and have a look at the messages? I don't have a Vitoconnect so I can't test how this device queries the machine (bulk, burst...). So we first create an ESP passthrough device and log the messages to MQTT?

PS To avoid any possible trouble, you should tie GPIO15 to ground to suppress boot messages.

**Comment by challo2018** [comment]
Currently Viessmann OptoLink Adapter (CP2102 Serial Chip) is connected to vitoconnect via USB.

I can build DIY Optolink adapter and connect it to Serial2 and connect ESP32 USB Port(Serial1) to the vitoconnect. 

We can start with your suggestion to have first passthrough and logging only.

**Comment by bertmelis** [comment]
**UNTESTED** (Quick and dirty, I don't even know it compiles or not.)
MQTT based on https://github.com/bertmelis/espMqttClient

```cpp
#include <Arduino.h>
#include <WiFi.h>
#include <cbuf.h>  // Circular buffer from Espressif Arduino framework

#include <espMqttclient.h>

#define WIFI_SSID "yourSSID"
#define WIFI_PASSWORD "yourpass"

#define MQTT_HOST IPAddress(192, 168, 1, 10)
#define MQTT_PORT 1883

#define USBSerial Serial
#define OptolinkSerial Serial1
#define BUFFERSIZE 256

cbuf txBuffer(BUFFERSIZE);
cbuf rxBuffer(BUFFERSIZE);
uint8_t buf[BUFFERSIZE];

void setup() {
  USBSerial.begin(4800, SERIAL_8E2);
  OptolinkSerial.begin(4800, SERIAL_8E2);
  
  mqttClient.setServer(config.hostname, config.port)
            .setServer(config.hostIP, config.port);

  WiFi.begin(WIFI_SSID, WIFI_PASSWORD);
  while (WiFi.status() != WL_CONNECTED) {
    delay(5);
  }
  mqttClient.connect();
}

void loop() {
  // 1. read data from USB into txBuffer
  int fromUSB = USBSerial.available();
  if (fromUSB > 0) {
    size_t room = txBuffer.room();
    USBSerial.read(buf, room);
    txBuffer.write(buf, room);
    mqttClient.publish("VitoWiFiPassthrough/fromUSB", 0, false, buf, room);
  }

  // 2. write txBuffer to Optolink
  while (txBuffer.available()) {
    if (OptolinkSerial.write(txBuffer.read()) < 1) break;
  }

   // 3. read data from Optolink into rxBuffer
  int fromoptolink = OptolinkSerial.available();
  if (fromoptolink > 0) {
    size_t room = txBuffer.room();
    OptolinkSerial.read(buf, room);
    rxBuffer.write(buf, room);
    mqttClient.publish("VitoWiFiPassthrough/toUSB", 0, false, buf, room);
  }

  // 4. write rxBuffer to USB
  while (rxBuffer.available()) {
    if (USBSerial.write(rxBuffer.read()) < 1) break;
  }
}

```

**Comment by bertmelis** [comment]
The above is (or should be) a simple USB <--> Serial forwarder.
Instead of publishing the passed-through message via MQTT, you could parse using `VitoWiFiInternals::ParserVS2` and it will give you a (reference to a) `VitoWiFi::PacketVS2`.

You can still choose to only parse messages coming from USB or only from Serial or both. Use a separate parser if you want both.
Depending on what the data looks like, the data is directly usable or has to be further split up.

If you could test the sample above and let me know how it goes...

**Comment by challo2018** [comment]
Thank you!! I will build up my DIY Optolink and check the code as quick as possible!

**Comment by bertmelis** [comment]
closing. feel free to reopen if further help is needed.

**Comment by bertmelis** [comment]
Maybe this is a solution for you: https://www.wemos.cc/en/latest/s3/s3.html

If Vitoconnect recognizes the ESP32-S3 as adapter, you could connect the original optolink to the ESP via the OTG port. Then this library could be integrated to communicate with the optolink: https://github.com/bertmelis/USBHostSerial

I have yet to adapt VitoWiFi to be compatible.

**Comment by challo2018** [comment]
i will get an ESP32-S3 and come back to you for test. Thank you!

**Comment by challo2018** [comment]
@bertmelis : I got ESP32-S3 for Testing your suggestion. Can you give me some hint on how to proceed with testing?

**Comment by alvv-z** [comment]
I would be willing to help test this. I have a [Waveshare ESP32-S3-Zero](https://www.waveshare.com/wiki/ESP32-S3-Zero) to which I have wired a DIY adapter (made of SFH309-FA and L-7104SF4BT). The plan is to use esphome (with, e.g., [esphome_vitoconnect](https://github.com/dannerph/esphome_vitoconnect)) so the vitoconnect can work in parallel to homeassistant.

**Comment by bertmelis** [comment]
You should connect the ESP to Vitoconnect and forward any messages to the DIY optolink and vice versa. The unknown for me is whether Vitoconnect detects the ESP or not. If Vitoconnect doesn't find suitable drivers it won't work, it can't communicate to the ESP. An external USB-to-UART chip (CP2102!) might be a solution then.

**Comment by alvv-z** [comment]
I'm not sure if an external chip would be needed, since the esp32-s3 natively support USB-to-UART IIRC.

**Reply by challo2018** [reply]
@alvv-z : Any news on your tests with esp32-s3?

**Reply by alvv-z** [reply]
Sadly, I haven't been able to test it out, since I don't often have physical access. It might be a while until I manage to test.

**Reply by idstein** [reply]
It works. I have a custom optolink bridge code in my esphome, but hopefully can soon give something back to VitoWiFi and ESPHome Optolink PRs to close the gap.

On a physical side the minimum required hardware is a CP2102 USB to UART bridge, because the VitoConnect OTPO 2 only supports this USB driver. For the smoothes adaption I use a ESP32-S3 with CP2102 to act as USB device to VitoConnect and utilise the built-in USB OTG host mode to re-use the existing Viessmann Optolink cable with its builtin CP2102 chipset.

Vitoconnect <-- USB CP2102 to serial UART --> ESP32-S3 <-- USB slave device CP2102 --> Viessmann Optolink Cable


The ESP32-S3 dev boards can draw enough current via the Vitoconnect, but the boards layout often do not forward the 5V VBUS to the USB slave port!
